src/main/java/com/microsoft/azure/relay/HybridHttpConnection.java

```diff
@@ -346,21 +354,46 @@ public void setWriteTimeout(Duration writeTimeout) {
 			return CompletableFuture.completedFuture(null);
 		}
 
+		/**
+		 * Writes the specified byte to this response stream.
+		 */
 		@Override
-		public void write(int b) {
-			this.writeAsync(new byte[] { (byte) b }, 0, 1).join();
+		public void write(int b) throws IOException {
+			this.write(new byte[] { (byte) b }, 0, 1);
 		}
 
-		public void write(byte[] bytes) {
-			this.writeAsync(bytes, 0, bytes.length).join();
+		/**
+		 * Writes b.length bytes from the specified byte array to this response stream.
+		 */
+		@Override
+		public void write(byte[] b) throws IOException {
+			this.write(b, 0, b.length);
+		}
+		
+		/**
+		 * Writes len bytes from the specified byte array starting at offset off to this response stream.
+		 */
+		@Override
+		public void write(byte[] b, int off, int len) throws IOException {
+			try {
+				this.writeAsync(b, off, len).join();
+			} catch (CompletionException e) {
+				throw new IOException(e.getCause());
+			}
 		}
 
-		public void write(String text) {
-			this.writeAsync(text.getBytes(StringUtil.UTF8), 0, text.length()).join();
+		/**
+		 * Writes the given text to this response stream.
+		 */
+		public void write(String text) throws IOException {
+			this.write(text.getBytes(StringUtil.UTF8), 0, text.length());
```

Comment:
>text.length() [](start = 49, length = 13)

Sorry I didn't catch this before.

text.length() gives the number of characters which may not be the same length as the number of bytes. If this overload is truly desired (I wonder if we could do without it) then it needs to capture the byte[] from text.getBytes() and use the length of that array. Also, I suppose it should be documented that this serializes a String using UTF-8 encoding.

Reply:
This may have existed before the PR any how and could be addressed in a future PR.

---
In reply to: [293595994](https://github.com/Azure/azure-relay-java/pull/46#discussion_r293595994) [](ancestors = 293595994)

Reply:
will remove this overload

---
In reply to: [293596279](https://github.com/Azure/azure-relay-java/pull/46#discussion_r293596279) [](ancestors = 293596279,293595994)